
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>AGW Tutorial &#8212; SCOT+</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=384b581d" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=efea14e4"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=36754332"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'AGW_tutorial';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Unbalanced Cell Type Alignment" href="pbmc_unbalanced.html" />
    <link rel="prev" title="Fused Formulation Tutorial" href="fused_tutorial.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/SCOTplus_logo.png" class="logo__image only-light" alt="SCOT+ - Home"/>
    <script>document.write(`<img src="_static/SCOTplus_logo.png" class="logo__image only-dark" alt="SCOT+ - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Welcome to SCOT+
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Core Tutorials</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="installation_instructions.html">Getting Set Up</a></li>
<li class="toctree-l1"><a class="reference internal" href="theory.html">The Evolution of SCOT+</a></li>
<li class="toctree-l1"><a class="reference internal" href="setup_tutorial.html">Setup Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="UGW_tutorial.html">UGW Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="UCOOT_tutorial.html">UCOOT Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="fused_tutorial.html">Fused Formulation Tutorial</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">AGW Tutorial</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Exploration on Real World Datasets</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="pbmc_unbalanced.html">Unbalanced Cell Type Alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="citeseq_sample_supervision.html">CITE-seq Sample Supervision and Hyperparameter Grid Search</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/scotplus/book_source.git" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/scotplus/book_source.git/issues/new?title=Issue%20on%20page%20%2FAGW_tutorial.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/AGW_tutorial.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>AGW Tutorial</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#preprocessing">Preprocessing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#default-agw">Default AGW</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#entropic-regularization">Entropic Regularization</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#marginal-relaxation">Marginal Relaxation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#alpha-and-balancing-gw-and-coot">Alpha and Balancing GW and COOT</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#supervision-extension-of-fused-formulation-to-agw">Supervision (Extension of Fused Formulation to AGW)</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="agw-tutorial">
<h1>AGW Tutorial<a class="headerlink" href="#agw-tutorial" title="Link to this heading">#</a></h1>
<p>In this tutorial, we will examine our newest tool, AGW. AGW blends together UGW and UCOOT to incentivize sample alignments that are informed by both feature correspondence and local geometry. We will again use the CITE-seq dataset from earlier tutorials, with the goal of recovering the underlying 1-1 alignment of the 1000 samples by considering expression and antibody data separately.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>If you have not yet configured a SCOT+ directory of some kind, see our installation instructions markdown. Once you complete those steps or set up your own virual environment, continue on here.</p>
<p>If you aren’t sure what any of the parameters for setting up a Solver object mean, try our setup tutorial for getting used to using the tool.</p>
<p>If you are looking for more detail on what the parameters of the alignment do in practice, start by visiting our UGW, UCOOT, and fused formulation tutorials. We will draw on all of these when examining AGW.</p>
<p>If you are unsure what some of the notation means throughout the rest of this document, try reading our optimal transport theory section to get more comfortable.</p>
<p>If you already know how to use AGW, continue on to the next chapter: our explorations of how AGW can be applied!</p>
</div>
<section id="preprocessing">
<h2>Preprocessing<a class="headerlink" href="#preprocessing" title="Link to this heading">#</a></h2>
<p>As usual, we set up PyTorch:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">torch</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Torch version: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">__version__</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;CUDA available: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">is_available</span><span class="p">()))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;CUDA version: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">version</span><span class="o">.</span><span class="n">cuda</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;CUDNN version: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">backends</span><span class="o">.</span><span class="n">cudnn</span><span class="o">.</span><span class="n">version</span><span class="p">()))</span>

<span class="n">use_cuda</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">is_available</span><span class="p">()</span>
<span class="n">device</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">device</span><span class="p">(</span><span class="s2">&quot;cuda:0&quot;</span> <span class="k">if</span> <span class="n">use_cuda</span> <span class="k">else</span> <span class="s2">&quot;cpu&quot;</span><span class="p">)</span>
<span class="n">torch</span><span class="o">.</span><span class="n">backends</span><span class="o">.</span><span class="n">cudnn</span><span class="o">.</span><span class="n">benchmark</span><span class="o">=</span><span class="kc">True</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Torch version: 2.1.0
CUDA available: False
CUDA version: None
CUDNN version: None
</pre></div>
</div>
</div>
</div>
<p>To begin, we will load in our preprocessed data. If you would like to preprocess the data in your own way, you can download the raw files for this dataset by running “sh download_scripts/CITEseq_download.sh” from the root of your SCOT+ directory. They will be loaded into the same folder as the preprocessed datasets.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%capture</span>
<span class="kn">from</span> <span class="nn">scotplus.solvers</span> <span class="kn">import</span> <span class="n">SinkhornSolver</span>
<span class="kn">from</span> <span class="nn">scotplus.utils.alignment</span> <span class="kn">import</span> <span class="n">compute_graph_distances</span><span class="p">,</span> <span class="n">get_barycentre</span><span class="p">,</span> <span class="n">FOSCTTM</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">umap</span>
<span class="kn">from</span> <span class="nn">sklearn.preprocessing</span> <span class="kn">import</span> <span class="n">normalize</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;font.family&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Helvetica Neue&#39;</span>

<span class="n">adt_raw</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;./data/CITEseq/citeseq_adt_normalized_1000cells.csv&quot;</span><span class="p">)</span>
<span class="n">rna_raw</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;./data/CITEseq/citeseq_rna_normalizedFC_1000cells.csv&quot;</span><span class="p">)</span>
<span class="n">adt_feat_labels</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CD11a&quot;</span><span class="p">,</span><span class="s2">&quot;CD11c&quot;</span><span class="p">,</span><span class="s2">&quot;CD123&quot;</span><span class="p">,</span><span class="s2">&quot;CD127-IL7Ra&quot;</span><span class="p">,</span><span class="s2">&quot;CD14&quot;</span><span class="p">,</span><span class="s2">&quot;CD16&quot;</span><span class="p">,</span><span class="s2">&quot;CD161&quot;</span><span class="p">,</span><span class="s2">&quot;CD19&quot;</span><span class="p">,</span><span class="s2">&quot;CD197-CCR7&quot;</span><span class="p">,</span><span class="s2">&quot;CD25&quot;</span><span class="p">,</span><span class="s2">&quot;CD27&quot;</span><span class="p">,</span><span class="s2">&quot;CD278-ICOS&quot;</span><span class="p">,</span><span class="s2">&quot;CD28&quot;</span><span class="p">,</span><span class="s2">&quot;CD3&quot;</span><span class="p">,</span><span class="s2">&quot;CD34&quot;</span><span class="p">,</span><span class="s2">&quot;CD38&quot;</span><span class="p">,</span><span class="s2">&quot;CD4&quot;</span><span class="p">,</span><span class="s2">&quot;CD45RA&quot;</span><span class="p">,</span><span class="s2">&quot;CD45RO&quot;</span><span class="p">,</span><span class="s2">&quot;CD56&quot;</span><span class="p">,</span><span class="s2">&quot;CD57&quot;</span><span class="p">,</span><span class="s2">&quot;CD69&quot;</span><span class="p">,</span><span class="s2">&quot;CD79b&quot;</span><span class="p">,</span><span class="s2">&quot;CD8a&quot;</span><span class="p">,</span><span class="s2">&quot;HLA.DR&quot;</span><span class="p">]</span>
<span class="n">rna_feat_labels</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;ITGAL&quot;</span><span class="p">,</span><span class="s2">&quot;ITGAX&quot;</span><span class="p">,</span><span class="s2">&quot;IL3RA&quot;</span><span class="p">,</span><span class="s2">&quot;IL7R&quot;</span><span class="p">,</span><span class="s2">&quot;CD14&quot;</span><span class="p">,</span><span class="s2">&quot;FCGR3A&quot;</span><span class="p">,</span><span class="s2">&quot;KLRB1&quot;</span><span class="p">,</span><span class="s2">&quot;CD19&quot;</span><span class="p">,</span><span class="s2">&quot;CCR7&quot;</span><span class="p">,</span><span class="s2">&quot;IL2RA&quot;</span><span class="p">,</span><span class="s2">&quot;CD27&quot;</span><span class="p">,</span><span class="s2">&quot;ICOS&quot;</span><span class="p">,</span><span class="s2">&quot;CD28&quot;</span><span class="p">,</span><span class="s2">&quot;CD3E&quot;</span><span class="p">,</span><span class="s2">&quot;CD34&quot;</span><span class="p">,</span><span class="s2">&quot;CD38&quot;</span><span class="p">,</span><span class="s2">&quot;CD4&quot;</span><span class="p">,</span><span class="s2">&quot;PTPRC&quot;</span><span class="p">,</span><span class="s2">&quot;PTPRC&quot;</span><span class="p">,</span><span class="s2">&quot;NCAM1&quot;</span><span class="p">,</span><span class="s2">&quot;B3GAT1&quot;</span><span class="p">,</span><span class="s2">&quot;CD69&quot;</span><span class="p">,</span><span class="s2">&quot;CD79B&quot;</span><span class="p">,</span><span class="s2">&quot;CD8A&quot;</span><span class="p">,</span><span class="s2">&quot;HLA-DRA&quot;</span><span class="p">]</span> 
<span class="n">samp_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Cell </span><span class="si">{0}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">adt_raw</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span>
</pre></div>
</div>
</div>
</div>
<p>From here, we can load our data into usable DataFrames, as in prior tutorials.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># l2 normalization of both datasets, which we found to help with single cell applications</span>
<span class="n">adt</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">normalize</span><span class="p">(</span><span class="n">adt_raw</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">transpose</span><span class="p">()))</span>
<span class="n">rna</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">normalize</span><span class="p">(</span><span class="n">rna_raw</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">transpose</span><span class="p">()))</span>

<span class="c1"># annotation of both domains</span>
<span class="n">adt</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">adt</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">samp_labels</span><span class="p">,</span> <span class="n">adt_feat_labels</span>
<span class="n">rna</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">rna</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">samp_labels</span><span class="p">,</span> <span class="n">rna_feat_labels</span>
</pre></div>
</div>
</div>
</div>
<p>In order to keep our objective in mind, let’s look at the UMAP/PCA of each domain pre-alignment one more time:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># we fit these objects now, and use them to transform our aligned data later on</span>
<span class="n">adt_um</span> <span class="o">=</span> <span class="n">umap</span><span class="o">.</span><span class="n">UMAP</span><span class="p">(</span><span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">rna_um</span> <span class="o">=</span> <span class="n">umap</span><span class="o">.</span><span class="n">UMAP</span><span class="p">(</span><span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="n">adt_um</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">adt</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())</span>
<span class="n">rna_um</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">rna</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())</span>

<span class="n">original_adt_um</span><span class="o">=</span><span class="n">adt_um</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">adt</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())</span>
<span class="n">original_rna_um</span><span class="o">=</span><span class="n">rna_um</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">rna</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())</span>

<span class="c1"># visualization of the global geometry</span>
<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span><span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">original_adt_um</span><span class="p">[:,</span><span class="mi">0</span><span class="p">],</span> <span class="n">original_adt_um</span><span class="p">[:,</span><span class="mi">1</span><span class="p">],</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;ADT Domain </span><span class="se">\n</span><span class="s2"> before Alignment&quot;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">original_rna_um</span><span class="p">[:,</span><span class="mi">0</span><span class="p">],</span> <span class="n">original_rna_um</span><span class="p">[:,</span><span class="mi">1</span><span class="p">],</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Gene Expression Domain </span><span class="se">\n</span><span class="s2"> before Alignment&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/ca94eab12bb6011bf8b3664b9e740184157664ef41ddfb3a4bc6488dbd5b47cc.png" src="_images/ca94eab12bb6011bf8b3664b9e740184157664ef41ddfb3a4bc6488dbd5b47cc.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">sklearn.decomposition</span> <span class="kn">import</span> <span class="n">PCA</span>

<span class="c1"># again, we fit these now so that we can transform our aligned data later on</span>
<span class="n">adt_pca</span><span class="o">=</span><span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">rna_pca</span><span class="o">=</span><span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

<span class="n">adt_pca</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">adt</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())</span>
<span class="n">rna_pca</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">rna</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())</span>

<span class="n">original_adt_pca</span><span class="o">=</span><span class="n">adt_pca</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">adt</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())</span>
<span class="n">original_rna_pca</span><span class="o">=</span><span class="n">rna_pca</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">rna</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())</span>

<span class="c1"># visualization of the global geometry</span>
<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span><span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">original_adt_pca</span><span class="p">[:,</span><span class="mi">0</span><span class="p">],</span> <span class="n">original_adt_pca</span><span class="p">[:,</span><span class="mi">1</span><span class="p">],</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;ADT Domain </span><span class="se">\n</span><span class="s2"> Before Alignment&quot;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">original_rna_pca</span><span class="p">[:,</span><span class="mi">0</span><span class="p">],</span> <span class="n">original_rna_pca</span><span class="p">[:,</span><span class="mi">1</span><span class="p">],</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Gene Expression Domain </span><span class="se">\n</span><span class="s2"> Before Alignment&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/cd9e995a62ffce9cbf1ea9d10b8b9f7ca823f886ab673d8309cdd8ed49e02e76.png" src="_images/cd9e995a62ffce9cbf1ea9d10b8b9f7ca823f886ab673d8309cdd8ed49e02e76.png" />
</div>
</div>
</section>
<section id="default-agw">
<h2>Default AGW<a class="headerlink" href="#default-agw" title="Link to this heading">#</a></h2>
<p>Next, we can instantiate a Solver object to begin aligning our data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scot</span> <span class="o">=</span> <span class="n">SinkhornSolver</span><span class="p">(</span><span class="n">nits_uot</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span> <span class="n">tol_uot</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>From here, we can begin by establishing pairwise distance matrices for our domains, as per the UGW tutorial:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># knn connectivity distance</span>
<span class="n">D_adt_knn</span><span class="p">,</span> <span class="n">D_rna_knn</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span><span class="n">compute_graph_distances</span><span class="p">(</span><span class="n">adt</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;connectivity&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">),</span> <span class="n">torch</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span><span class="n">compute_graph_distances</span><span class="p">(</span><span class="n">rna</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;connectivity&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now that we are all set up to run an alignment between <span class="math notranslate nohighlight">\(X\)</span> and <span class="math notranslate nohighlight">\(Y\)</span>, we will pause for a moment to examine what AGW is trying to do. As we did with UGW and UCOOT, let’s examine the objective function that AGW is hoping to minimize. First, recall our UGW objective function:</p>
<p><span class="math notranslate nohighlight">\(GW(\pi_s) = L_{GW}(D(x_i, x_j), D(y_k, y_l)) \cdot \pi_{s_{ij}} \cdot \pi_{s_{kl}}\)</span> for all samples <span class="math notranslate nohighlight">\(i, j\)</span> in <span class="math notranslate nohighlight">\(X\)</span> and <span class="math notranslate nohighlight">\(k, l\)</span> in <span class="math notranslate nohighlight">\(Y\)</span>. Recall that <span class="math notranslate nohighlight">\(L_{GW}\)</span> is squared euclidean distance for our purposes. Now, we re-examine the UCOOT objective function:</p>
<p><span class="math notranslate nohighlight">\(COOT(\pi_s, \pi_f) = L_{COOT}(X_{ij}, Y_{kl}) \cdot \pi_{s_{ik}} \cdot \pi_{f_{jl}}\)</span> for all samples <span class="math notranslate nohighlight">\(i\)</span> in <span class="math notranslate nohighlight">\(X\)</span>, features <span class="math notranslate nohighlight">\(j\)</span> in <span class="math notranslate nohighlight">\(X\)</span>, samples <span class="math notranslate nohighlight">\(k\)</span> in <span class="math notranslate nohighlight">\(Y\)</span>, and features <span class="math notranslate nohighlight">\(l\)</span> in <span class="math notranslate nohighlight">\(Y\)</span>. Again, <span class="math notranslate nohighlight">\(L_{COOT}\)</span> is squared euclidean distance for our purposes. Given these two functions, our AGW objective function is a convex combination as follows:</p>
<p><span class="math notranslate nohighlight">\(\alpha GW(\pi_s) + (1 - \alpha) COOT(\pi_s, \pi_f)\)</span></p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Note that we have a new hyperparameter <span class="math notranslate nohighlight">\(\alpha\)</span> – this is NOT the same <span class="math notranslate nohighlight">\(\beta\)</span>, which we called <span class="math notranslate nohighlight">\(\alpha\)</span> in prior versions of the tool. In this case, <span class="math notranslate nohighlight">\(\alpha\)</span> trades off how much we use GW versus COOT in our optimization procedure.</p>
</div>
<p>This AGW objective function allows us to continue employing BCD, considering that we can minimize the GW and COOT part of the costs iteratively. In this sense, each “block” consists of two blocks as we understood them in the UCOOT and UGW tutorials. A new parameter we can specify is nits_gw, which determines how many times we run a GW block (back and forth optimization of <span class="math notranslate nohighlight">\(\pi_s\)</span> and <span class="math notranslate nohighlight">\(\pi_s'\)</span>, our copy of <span class="math notranslate nohighlight">\(\pi_s\)</span> from GW) per COOT block (back and for optimization of <span class="math notranslate nohighlight">\(\pi_s\)</span> and <span class="math notranslate nohighlight">\(\pi_f\)</span>)</p>
<p>AGW allows us to look at local geometry (GW) and feature correspondence (COOT) when optimizing a given <span class="math notranslate nohighlight">\(\pi_s\)</span>. If you would like more information on the theory behind why AGW works, look at our theory document early in this tutorial chapter. Let’s try an alignment:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pi_samp</span><span class="p">,</span> <span class="n">pi_samp_prime</span><span class="p">,</span> <span class="n">pi_feat</span> <span class="o">=</span> <span class="n">scot</span><span class="o">.</span><span class="n">agw</span><span class="p">(</span><span class="n">rna</span><span class="p">,</span> <span class="n">adt</span><span class="p">,</span> <span class="n">D_rna_knn</span><span class="p">,</span> <span class="n">D_adt_knn</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">eps</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We can immediately visualize the two key resulting coupling matrices. The process for scoring and visualizing this alignment is the same as in prior notebooks.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<span class="n">_</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>

<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pi_feat</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">rna</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">adt</span><span class="o">.</span><span class="n">columns</span><span class="p">),</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Blues&#39;</span><span class="p">,</span> <span class="n">square</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">pi_samp</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Blues&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/ffe788952d7bf33beeb95d994389477e776b67f6a09bd8bfc129dc44eb90ebd1.png" src="_images/ffe788952d7bf33beeb95d994389477e776b67f6a09bd8bfc129dc44eb90ebd1.png" />
</div>
</div>
</section>
<section id="entropic-regularization">
<h2>Entropic Regularization<a class="headerlink" href="#entropic-regularization" title="Link to this heading">#</a></h2>
<p>Entropic regularization functions the same as in prior tutorials – if you are curious about what entropic regularization does, look at the UGW or UCOOT tutorials. We can quickly show that this similarity holds by experimenting with a few different values of epsilon.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pi_samp_sm</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">pi_feat_sm</span> <span class="o">=</span> <span class="n">scot</span><span class="o">.</span><span class="n">agw</span><span class="p">(</span><span class="n">rna</span><span class="p">,</span> <span class="n">adt</span><span class="p">,</span> <span class="n">D_rna_knn</span><span class="p">,</span> <span class="n">D_adt_knn</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">eps</span><span class="o">=</span><span class="mf">5e-4</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">pi_samp_med</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">pi_feat_med</span> <span class="o">=</span> <span class="n">scot</span><span class="o">.</span><span class="n">agw</span><span class="p">(</span><span class="n">rna</span><span class="p">,</span> <span class="n">adt</span><span class="p">,</span> <span class="n">D_rna_knn</span><span class="p">,</span> <span class="n">D_adt_knn</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">eps</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">pi_samp_lg</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">pi_feat_lg</span> <span class="o">=</span> <span class="n">scot</span><span class="o">.</span><span class="n">agw</span><span class="p">(</span><span class="n">rna</span><span class="p">,</span> <span class="n">adt</span><span class="p">,</span> <span class="n">D_rna_knn</span><span class="p">,</span> <span class="n">D_adt_knn</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">eps</span><span class="o">=</span><span class="mf">1e-1</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">pi_feat</span><span class="p">,</span> <span class="n">pi_samp</span><span class="p">,</span> <span class="n">size</span> <span class="ow">in</span> <span class="p">[(</span><span class="n">pi_feat_sm</span><span class="p">,</span> <span class="n">pi_samp_sm</span><span class="p">,</span> <span class="s1">&#39;eps=5e-4&#39;</span><span class="p">),</span> <span class="p">(</span><span class="n">pi_feat_med</span><span class="p">,</span> <span class="n">pi_samp_med</span><span class="p">,</span> <span class="s1">&#39;eps=1e-3&#39;</span><span class="p">),</span> <span class="p">(</span><span class="n">pi_feat_lg</span><span class="p">,</span> <span class="n">pi_samp_lg</span><span class="p">,</span> <span class="s1">&#39;eps=1e-1&#39;</span><span class="p">)]:</span>
    
    <span class="n">_</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>

    <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pi_feat</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">rna</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">adt</span><span class="o">.</span><span class="n">columns</span><span class="p">),</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Blues&#39;</span><span class="p">,</span> <span class="n">square</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">pi_samp</span><span class="p">[:</span><span class="mi">100</span><span class="p">,:</span><span class="mi">100</span><span class="p">],</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Blues&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/99bcca07f8e822d1937e60d7addbf8813615997da9cc9e30fb29b1e3163f141c.png" src="_images/99bcca07f8e822d1937e60d7addbf8813615997da9cc9e30fb29b1e3163f141c.png" />
<img alt="_images/842d666fb0468c5f269253a62480bb25deaca4d1705fba88b23049d76fd55c01.png" src="_images/842d666fb0468c5f269253a62480bb25deaca4d1705fba88b23049d76fd55c01.png" />
<img alt="_images/a91ced0299499a12e05f4d9a8bfd6cbce00248655db98009766d588f52b38c48.png" src="_images/a91ced0299499a12e05f4d9a8bfd6cbce00248655db98009766d588f52b38c48.png" />
</div>
</div>
<p>As seen above, high epsilon still encourages density, while a low epsilon leads to more sparse coupling matrices. As seen above, the low epsilon badly “overfits”, as a reminder that we want to strike a balance.</p>
</section>
<section id="marginal-relaxation">
<h2>Marginal Relaxation<a class="headerlink" href="#marginal-relaxation" title="Link to this heading">#</a></h2>
<p>Unlike UGW and UCOOT, our current version of AGW does not allow for marginal relaxation (although, since our backend solver is fully generalized, you can experiment with the notion of it). We may add <span class="math notranslate nohighlight">\(\rho\)</span> as a hyperparameter in the future, but for the moment, we do not support it. The lack of marginal relaxation in AGW places more emphasis on sample supervision and the initial distributions we use.</p>
</section>
<section id="alpha-and-balancing-gw-and-coot">
<h2>Alpha and Balancing GW and COOT<a class="headerlink" href="#alpha-and-balancing-gw-and-coot" title="Link to this heading">#</a></h2>
<p>Now, we can examine AGW closely. We will start by looking at <span class="math notranslate nohighlight">\(\alpha\)</span>, and then move on to algorithmic decisions and sample/feature supervision.</p>
<p>To start, we will compare the <span class="math notranslate nohighlight">\(\alpha = 0\)</span> and <span class="math notranslate nohighlight">\(\alpha = 1\)</span> cases to UGW and UCOOT, using the Sinkhorn algorithm (the common algorithm that both AGW and UGW/UCOOT use).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># changing entropic_mode to mimic GW-like regularization</span>
<span class="n">pi_samp_ngw</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">scot</span><span class="o">.</span><span class="n">agw</span><span class="p">(</span><span class="n">rna</span><span class="p">,</span> <span class="n">adt</span><span class="p">,</span> <span class="n">D_rna_knn</span><span class="p">,</span> <span class="n">D_adt_knn</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">eps</span><span class="o">=</span><span class="mf">3e-3</span><span class="p">,</span> <span class="n">entropic_mode</span><span class="o">=</span><span class="s1">&#39;independent&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">pi_samp_ogw</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">scot</span><span class="o">.</span><span class="n">gw</span><span class="p">(</span><span class="n">D_rna_knn</span><span class="p">,</span> <span class="n">D_adt_knn</span><span class="p">,</span> <span class="n">eps</span><span class="o">=</span><span class="mf">3e-3</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">pi_samp_ncoot</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">pi_feat_ncoot</span> <span class="o">=</span> <span class="n">scot</span><span class="o">.</span><span class="n">agw</span><span class="p">(</span><span class="n">rna</span><span class="p">,</span> <span class="n">adt</span><span class="p">,</span> <span class="n">D_rna_knn</span><span class="p">,</span> <span class="n">D_adt_knn</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">eps</span><span class="o">=</span><span class="mf">3e-3</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">pi_samp_ocoot</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">pi_feat_ocoot</span> <span class="o">=</span> <span class="n">scot</span><span class="o">.</span><span class="n">coot</span><span class="p">(</span><span class="n">rna</span><span class="p">,</span> <span class="n">adt</span><span class="p">,</span> <span class="n">eps</span><span class="o">=</span><span class="mf">3e-3</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We can check whether these coupling matrices are similar, as we would expect:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># checking gw vs. alpha = 1</span>
<span class="n">torch</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">pi_samp_ngw</span><span class="p">),</span> <span class="n">torch</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">pi_samp_ogw</span><span class="p">),</span> <span class="n">torch</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">pi_samp_ngw</span> <span class="o">-</span> <span class="n">pi_samp_ogw</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(tensor(1.), tensor(1.0000), tensor(0.2599))
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># checking coot vs. alpha = 0</span>
<span class="n">torch</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">pi_samp_ncoot</span><span class="p">),</span> <span class="n">torch</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">pi_samp_ocoot</span><span class="p">),</span> <span class="n">torch</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">pi_samp_ncoot</span> <span class="o">-</span> <span class="n">pi_samp_ocoot</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(tensor(1.0000), tensor(1.0000), tensor(1.2190e-06))
</pre></div>
</div>
</div>
</div>
<p>As we can see above, we get nearly identical reults (to be expected). From here, we can look at an intermediate value of <span class="math notranslate nohighlight">\(\alpha\)</span>, and compare visualizations with <span class="math notranslate nohighlight">\(\alpha=0\)</span> and <span class="math notranslate nohighlight">\(\alpha=1\)</span>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pi_samp_half</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">pi_feat_half</span> <span class="o">=</span> <span class="n">scot</span><span class="o">.</span><span class="n">agw</span><span class="p">(</span><span class="n">rna</span><span class="p">,</span> <span class="n">adt</span><span class="p">,</span> <span class="n">D_rna_knn</span><span class="p">,</span> <span class="n">D_adt_knn</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">eps</span><span class="o">=</span><span class="mf">3e-3</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We begin by projecting and scoring these alignments:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">aligned_rna_ngw</span> <span class="o">=</span> <span class="n">get_barycentre</span><span class="p">(</span><span class="n">adt</span><span class="p">,</span> <span class="n">pi_samp_ngw</span><span class="p">)</span>
<span class="n">aligned_rna_half</span> <span class="o">=</span> <span class="n">get_barycentre</span><span class="p">(</span><span class="n">adt</span><span class="p">,</span> <span class="n">pi_samp_half</span><span class="p">)</span>
<span class="n">aligned_rna_ncoot</span> <span class="o">=</span> <span class="n">get_barycentre</span><span class="p">(</span><span class="n">adt</span><span class="p">,</span> <span class="n">pi_samp_ncoot</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">aligned_rna</span><span class="p">,</span> <span class="n">size</span> <span class="ow">in</span> <span class="p">[(</span><span class="n">aligned_rna_ncoot</span><span class="p">,</span> <span class="s1">&#39;alpha=0&#39;</span><span class="p">),</span> <span class="p">(</span><span class="n">aligned_rna_half</span><span class="p">,</span> <span class="s1">&#39;alpha=0.5&#39;</span><span class="p">),</span> <span class="p">(</span><span class="n">aligned_rna_ngw</span><span class="p">,</span> <span class="s1">&#39;alpha=1&#39;</span><span class="p">)]:</span>
    <span class="n">fracs</span> <span class="o">=</span> <span class="n">FOSCTTM</span><span class="p">(</span><span class="n">adt</span><span class="p">,</span> <span class="n">aligned_rna</span><span class="o">.</span><span class="n">numpy</span><span class="p">())</span>
    <span class="n">legend_label</span><span class="o">=</span><span class="s2">&quot;AGW alignment FOSCTTM </span><span class="se">\n</span><span class="s2"> average value for </span><span class="si">{0}</span><span class="s2">: &quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">size</span><span class="p">)</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">fracs</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">fracs</span><span class="p">)),</span> <span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">fracs</span><span class="p">),</span> <span class="s2">&quot;b--&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">legend_label</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Cells&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Sorted FOSCTTM&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/64e26122caf5ff8863270985b6ed3b5d03afa65e3b5a7e9dff5a0aff3ddf9052.png" src="_images/64e26122caf5ff8863270985b6ed3b5d03afa65e3b5a7e9dff5a0aff3ddf9052.png" />
<img alt="_images/7208273891a60db2450d646b449087b04f2aa229e735c58a5e6d0ab700de7949.png" src="_images/7208273891a60db2450d646b449087b04f2aa229e735c58a5e6d0ab700de7949.png" />
<img alt="_images/757af862d5668b6a73281d0f3c3987feefd33039f283b1a7f040f19387a953de.png" src="_images/757af862d5668b6a73281d0f3c3987feefd33039f283b1a7f040f19387a953de.png" />
</div>
</div>
<p>In addition, we can examine the coupling matrices:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">pi_feat</span><span class="p">,</span> <span class="n">pi_samp</span><span class="p">,</span> <span class="n">size</span> <span class="ow">in</span> <span class="p">[(</span><span class="n">pi_feat_ncoot</span><span class="p">,</span> <span class="n">pi_samp_ncoot</span><span class="p">,</span> <span class="s1">&#39;alpha=0&#39;</span><span class="p">),</span> <span class="p">(</span><span class="n">pi_feat_half</span><span class="p">,</span> <span class="n">pi_samp_half</span><span class="p">,</span> <span class="s1">&#39;alpha=0.5&#39;</span><span class="p">),</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="mi">25</span><span class="p">,</span> <span class="mi">25</span><span class="p">))</span><span class="o">/</span><span class="p">(</span><span class="mi">25</span><span class="o">**</span><span class="mi">2</span><span class="p">),</span> <span class="n">pi_samp_ngw</span><span class="p">,</span> <span class="s1">&#39;alpha=1&#39;</span><span class="p">)]:</span>
    
    <span class="n">_</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>

    <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pi_feat</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">rna</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">adt</span><span class="o">.</span><span class="n">columns</span><span class="p">),</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Blues&#39;</span><span class="p">,</span> <span class="n">square</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    
    <span class="c1"># looking at a corner to get a sense</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">pi_samp</span><span class="p">[:</span><span class="mi">50</span><span class="p">,</span> <span class="p">:</span><span class="mi">50</span><span class="p">],</span><span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Blues&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/330d91704cc265c09057b32fccdc2a85be193c3552cb76bf5a6f43b5165d0ec3.png" src="_images/330d91704cc265c09057b32fccdc2a85be193c3552cb76bf5a6f43b5165d0ec3.png" />
<img alt="_images/302302f85affae7300f5f6a965efc3de7312db47bf3d655c0302c247ff412a6c.png" src="_images/302302f85affae7300f5f6a965efc3de7312db47bf3d655c0302c247ff412a6c.png" />
<img alt="_images/bf489045a586578655ee1f3427253a7034517c1ae46622772737f7cf03cc7d0b.png" src="_images/bf489045a586578655ee1f3427253a7034517c1ae46622772737f7cf03cc7d0b.png" />
</div>
</div>
<p>Note that the middle matrices, <span class="math notranslate nohighlight">\(\alpha=0.5\)</span>, appear to be somewhere in between the equivalent matrices for <span class="math notranslate nohighlight">\(\alpha=0\)</span> and <span class="math notranslate nohighlight">\(\alpha=1\)</span>. We recommend linearly searching over <span class="math notranslate nohighlight">\(\alpha\)</span>, as some datasets benefit more from focus on local geometry (high alpha), while some benefit from more focus on feature correspondence (low alpha, when features have strong relationships). As an example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># an example of alpha search</span>
<span class="n">pi_samp_dt</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">pi_feat_dt</span> <span class="o">=</span> <span class="p">{}</span>
<span class="c1"># for val in np.linspace(start=0, stop=-1, num=10):</span>
    <span class="c1"># pi_samp_dt[val], _, pi_feat_dt[val] = scot.agw(rna, adt, D_rna_knn, D_adt_knn, alpha=val, eps=eps)</span>
</pre></div>
</div>
</div>
</div>
<p>We generally find that smaller values of alpha (closer to COOT) yield the best performance.</p>
</section>
<section id="supervision-extension-of-fused-formulation-to-agw">
<h2>Supervision (Extension of Fused Formulation to AGW)<a class="headerlink" href="#supervision-extension-of-fused-formulation-to-agw" title="Link to this heading">#</a></h2>
<p>AGW allows for supervision in a similar way to UCOOT/UGW; whenever we calculate the local cost of an individual OT problem with respect to some matrix <span class="math notranslate nohighlight">\(\pi\)</span> that is subject to supervision via a <span class="math notranslate nohighlight">\(\beta \langle D, \pi \rangle\)</span> term, we add <span class="math notranslate nohighlight">\(\beta*D\)</span> to the local cost matrix. In particular, we can input our supervision to D = (D_samp, D_feat) as usual. In order to examine how supervision can help, let’s try supervising on the feature level and seeing how the scores improve.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>AGW uses the same beta parameter we had in UCOOT/UGW supervision. Always take a look at your local cost matrices before assigning beta, as your supervision should be relative to the magnitude of the costs the algorithm is applying.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">D_feat</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">identity</span><span class="p">(</span><span class="mi">25</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;float32&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="mi">25</span><span class="p">,</span> <span class="mi">25</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;float32&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>

<span class="n">pi_samp_unsup</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">pi_feat_unsup</span> <span class="o">=</span> <span class="n">scot</span><span class="o">.</span><span class="n">agw</span><span class="p">(</span><span class="n">rna</span><span class="p">,</span> <span class="n">adt</span><span class="p">,</span> <span class="n">D_rna_knn</span><span class="p">,</span> <span class="n">D_adt_knn</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">eps</span><span class="o">=</span><span class="mf">2e-3</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">pi_samp_medsup</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">pi_feat_medsup</span> <span class="o">=</span> <span class="n">scot</span><span class="o">.</span><span class="n">agw</span><span class="p">(</span><span class="n">rna</span><span class="p">,</span> <span class="n">adt</span><span class="p">,</span> <span class="n">D_rna_knn</span><span class="p">,</span> <span class="n">D_adt_knn</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">eps</span><span class="o">=</span><span class="mf">2e-3</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1e-1</span><span class="p">),</span> <span class="n">D</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">D_feat</span><span class="p">),</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">pi_samp_fullsup</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">pi_feat_fullsup</span> <span class="o">=</span> <span class="n">scot</span><span class="o">.</span><span class="n">agw</span><span class="p">(</span><span class="n">rna</span><span class="p">,</span> <span class="n">adt</span><span class="p">,</span> <span class="n">D_rna_knn</span><span class="p">,</span> <span class="n">D_adt_knn</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">eps</span><span class="o">=</span><span class="mf">2e-3</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">D</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">D_feat</span><span class="p">),</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">aligned_rna_unsup</span> <span class="o">=</span> <span class="n">get_barycentre</span><span class="p">(</span><span class="n">adt</span><span class="p">,</span> <span class="n">pi_samp_unsup</span><span class="p">)</span>
<span class="n">aligned_rna_medsup</span> <span class="o">=</span> <span class="n">get_barycentre</span><span class="p">(</span><span class="n">adt</span><span class="p">,</span> <span class="n">pi_samp_medsup</span><span class="p">)</span>
<span class="n">aligned_rna_fullsup</span> <span class="o">=</span> <span class="n">get_barycentre</span><span class="p">(</span><span class="n">adt</span><span class="p">,</span> <span class="n">pi_samp_fullsup</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We can score and visualize:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">aligned_rna</span><span class="p">,</span> <span class="n">mode</span> <span class="ow">in</span> <span class="p">[(</span><span class="n">aligned_rna_unsup</span><span class="p">,</span> <span class="s1">&#39;unsupervised&#39;</span><span class="p">),</span> <span class="p">(</span><span class="n">aligned_rna_medsup</span><span class="p">,</span> <span class="s1">&#39;some supervision&#39;</span><span class="p">),</span> <span class="p">(</span><span class="n">aligned_rna_fullsup</span><span class="p">,</span> <span class="s1">&#39;full supervision&#39;</span><span class="p">)]:</span>
    <span class="n">fracs</span> <span class="o">=</span> <span class="n">FOSCTTM</span><span class="p">(</span><span class="n">adt</span><span class="p">,</span> <span class="n">aligned_rna</span><span class="o">.</span><span class="n">numpy</span><span class="p">())</span>
    <span class="n">legend_label</span><span class="o">=</span><span class="s2">&quot;AGW alignment FOSCTTM </span><span class="se">\n</span><span class="s2"> average value for </span><span class="si">{0}</span><span class="s2">: &quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mode</span><span class="p">)</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">fracs</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">fracs</span><span class="p">)),</span> <span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">fracs</span><span class="p">),</span> <span class="s2">&quot;b--&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">legend_label</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Cells&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Sorted FOSCTTM&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/64f015c222143b2a63b0448cc37f81568c5651d0997b8cdad8870cbab1b930ac.png" src="_images/64f015c222143b2a63b0448cc37f81568c5651d0997b8cdad8870cbab1b930ac.png" />
<img alt="_images/6ece96a959e2e20e2cf5a29d56356dab23d41eda9c5665c234dd8caf82720f12.png" src="_images/6ece96a959e2e20e2cf5a29d56356dab23d41eda9c5665c234dd8caf82720f12.png" />
<img alt="_images/a479262fe9502f634827d7fd1264524570b05f3b3c3a25d53c9d3e125c2f816c.png" src="_images/a479262fe9502f634827d7fd1264524570b05f3b3c3a25d53c9d3e125c2f816c.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">pi_feat</span><span class="p">,</span> <span class="n">pi_samp</span><span class="p">,</span> <span class="n">mode</span> <span class="ow">in</span> <span class="p">[(</span><span class="n">pi_feat_unsup</span><span class="p">,</span> <span class="n">pi_samp_unsup</span><span class="p">,</span> <span class="s1">&#39;unsupervised&#39;</span><span class="p">),</span> <span class="p">(</span><span class="n">pi_feat_medsup</span><span class="p">,</span> <span class="n">pi_samp_medsup</span><span class="p">,</span> <span class="s1">&#39;some supervision&#39;</span><span class="p">),</span> <span class="p">(</span><span class="n">pi_feat_fullsup</span><span class="p">,</span> <span class="n">pi_samp_fullsup</span><span class="p">,</span> <span class="s1">&#39;full supervision&#39;</span><span class="p">)]:</span>
    
    <span class="n">_</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>

    <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pi_feat</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">rna</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">adt</span><span class="o">.</span><span class="n">columns</span><span class="p">),</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Blues&#39;</span><span class="p">,</span> <span class="n">square</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">ax2</span><span class="o">.</span><span class="n">spy</span><span class="p">(</span><span class="n">pi_samp</span><span class="p">,</span> <span class="n">precision</span><span class="o">=</span><span class="mf">0.0001</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/c1ae2527b82f20e2fd36886b096b3e5925b1d8a52da0b3101deda5b14cd68566.png" src="_images/c1ae2527b82f20e2fd36886b096b3e5925b1d8a52da0b3101deda5b14cd68566.png" />
<img alt="_images/a709056fca44b5d35cdc189920b29187e4d33aa348bd417d2ff7fb3514f166bb.png" src="_images/a709056fca44b5d35cdc189920b29187e4d33aa348bd417d2ff7fb3514f166bb.png" />
<img alt="_images/6839934a3613975c87c67478297987836cbee3875d7fe0ef74a7c037f8ce326b.png" src="_images/6839934a3613975c87c67478297987836cbee3875d7fe0ef74a7c037f8ce326b.png" />
</div>
</div>
<p>As we can see, feature level supervision greatly aids sample level alignment, as expected!</p>
<p>In one of our future application notebooks, we will examine potential ways to generate sample and feature supervision matrices given cell types (sample supervision) and prior knowledge (feature supervision).</p>
<p>We have now concluded the AGW tutorial. The key differences between AGW and UCOOT/UGW are:</p>
<ol class="arabic simple">
<li><p>AGW uses <span class="math notranslate nohighlight">\(\alpha\)</span> to denote the GW/COOT balance, rather than a supervision coefficient.</p></li>
<li><p>AGW does not allow for marginal relaxation at the moment.</p></li>
<li><p>AGW brings together GW/COOT simultaneously to improve alignment quality and account for both feature correspondence as well as local geometry.</p></li>
</ol>
<p>Now that you have reached this stage, we recommend viewing our application notebooks, which are still rolling out! These tutorials have attempted to help give you intuition for the parameters, algorithms, etc., but they do not display well what to do in practice on unpaired datasets and how to score these circumstances. Continue on to learn more about how AGW can better align your data.</p>
<p>Citations:</p>
<p>Pinar Demetci, Quang Huy Tran, Ievgen Redko, Ritambhara Singh. Revisiting invariances and introducing priors in Gromov-Wasserstein distances. arXiv, <a class="reference external" href="http://stat.ML">stat.ML</a>, 2023.</p>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "scotplus"
        },
        kernelOptions: {
            name: "scotplus",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'scotplus'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="fused_tutorial.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Fused Formulation Tutorial</p>
      </div>
    </a>
    <a class="right-next"
       href="pbmc_unbalanced.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Unbalanced Cell Type Alignment</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#preprocessing">Preprocessing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#default-agw">Default AGW</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#entropic-regularization">Entropic Regularization</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#marginal-relaxation">Marginal Relaxation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#alpha-and-balancing-gw-and-coot">Alpha and Balancing GW and COOT</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#supervision-extension-of-fused-formulation-to-agw">Supervision (Extension of Fused Formulation to AGW)</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By C. Baker
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>